replica/compaction_group.hh

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright (C) 2022-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+#include "database_fwd.hh"
+#include "sstables/sstable_set.hh"
+
+#pragma once
+
+namespace replica {
+
+// Compaction group is a set of SSTables which are eligible to be compacted together.
+// By this definition, we can say:
+//      - A group contains SSTables from the same table and shard
+//      - Each group can be thought of an isolated LSM tree.
+//      - Time Window strategy has one group per window
+class compaction_group {
+    table& _t;
+    std::unique_ptr<table::table_state> _table_state;
+    // Holds list of memtables for this group
+    lw_shared_ptr<memtable_list> _memtables;
+    // SSTable set which contains all non-maintenance sstables
+    lw_shared_ptr<sstables::sstable_set> _main_sstables;
+    // Holds SSTables created by maintenance operations, which need reshaping before integration into the main set
+    lw_shared_ptr<sstables::sstable_set> _maintenance_sstables;
+public:
+    compaction_group(table& t);
+
+    // Will stop ongoing compaction on behalf of this group, etc.
+    future<> stop();
```

Comment:
noexcept please

Reply:
So we can/should make flush noexcept too, for exactly that reason.
It should be a very rare reason to allow an async function to fail in 2 ways - throwing an exception, and returning a failed future, because this can double the error handling efforts (in the outermost calls.  if a function throws in a continuation we convert that into an exceptional future).

The most common technique to use is:
```c++
future<> compaction_group::stop() {
  try {
    return _memtables->flush().finally([this] {
        return _t._compaction_manager.remove(as_table_state());
    });
  } catch (...) {
    return current_exception_as_future<>();
  }
}
```

This will not terminate.

Another way to do it is:
`futurize_invoke([this] { return _memtables->flush(); }).finally( ... )`

but it's cumbersome and ugly in this case.

So I'd vote for making flush noexcept to start with, possibly catching any exceptions there as described above.

Reply:
That's orthogonal to this series, I think. I think I can start with your suggestion to convert any thrown exception into exception future and add a FIXME for the subsequent work.

Reply:
That = make flush noexcept.